src/statmodels.jl

```diff
@@ -39,6 +39,35 @@ the likelihood of the model.
 """
 deviance(obj::StatisticalModel) = error("deviance is not defined for $(typeof(obj)).")
 
+"""
+    hessian(obj::StatisticalModel)
+
+Return the Hessian of the statistical model.
+"""
+hessian(obj::StatisticalModel) = error("hessian is not defined for $(typeof(obj)).")
```

Comment:
[Hessian](https://en.wikipedia.org/wiki/Hessian_matrix) is a much more general term. In fact, several other libraries already define `hessian` functions (e.g. Calculus and ForwardDiff)

Reply:
This is true. This might be a case where it is justified to use `model_hessian`? A big part would be to provide a nice way for statistical models to play well with optimization packages.

Reply:
Is that really a problem? As you noted, several packages already use the same name and that's not the end of the world. It's a case where automatic merging of methods would make sense.

Reply:
Perhaps don't export it then?

Reply:
It's not a problem per se to export functions with conflicting names, the error only happens if somebody tries to call it while conflicting packages are loaded. If we don't export it from any of the potentially conflicting packages, we're end up in a strictly worse situation where you always need to fully qualify the function.

Reply:
Calculus exports `hessian` and ForwardDiff doesn't. I think it would be best at least for now to go with `model_hessian` to avoid any potential conflicting namespace with other optimization packages.

Reply:
If we're trying to get rid of these `model_*` functions, better not introduce new ones now.

Reply:
I think the idea was to remove it from function that do not need it (e.g., `model_response` to `response` or `model_weights` to `weights`). For functions that justify it, it should be fine. For example, `modelmatrix` is justified as `matrix` would be bad. I would have liked `hessian`, but sadly it is expected to be used in optimization packages and these not all require `StatsBase`.

As an aside, should `weights(obj::AbstractVector{T}) where T <: Real` be depreciated to just `Weights`? It seems like the Julian thing would be to keep constructors with uppercase. `pweights`, `aweights`, and the others seem just like short-hand for long to type structs.

Reply:
Yeah, but "model matrix" is a standard term in the field, probably because "matrix" is indeed too broad. "Model hessian" doesn't exist AFAIK.

Regarding `weights`, that pattern of lower case convenience constructors is used elsewhere in Julia, for example in Base.Iterators. Anyway, better discuss that elsewhere.

Reply:
What about skipping the `hessian` and implementing `eim` `oim` for the Expected Information Matrix and Observed Information Matrix? These are specific to statistical models (components for the `vcov`).

Reply:
Not sure. I kind of like the idea of using the same name `hessian` across packages given that it has a consistent meaning. We can avoid exporting it if people are concerned about that.

What do other software (languages, packages...) do?

Cc: @dmbates

Reply:
I prefer slightly longer names where possible, so how about `fisherinf` and `observedinf`?